src/par/FixColumnPartitions.c

```diff
@@ -19,138 +19,162 @@
-    for (i = 0; i < nrf; i++) {
-      rf = List_Entry(rfaces,i);
-      MF_Coords(rf,&nfv,fxyz);
-
-      fcen[0] = fcen[1] = fcen[2] = 0.0;
-      for (j = 0; j < nfv; j++) {
-        for (k = 0; k < 3; k++)
-          fcen[k] += fxyz[j][k];
-      }
-      for (k = 0; k < 3; k++)
-        fcen[k] /= nfv;
-
-      for (k = 0; k < 3; k++)
-        rfvec[k] = fcen[k]-rcen[k];
-      norm = sqrt(rfvec[0]*rfvec[0]+rfvec[1]*rfvec[1]+rfvec[2]*rfvec[2]);
-      for (k = 0; k < 3; k++)
-        rfvec[k] /= norm;
-
-      /* rfvec[2] is same as rfvec dotted with [0,0,1] */
-      if (rfvec[2] > maxdp) {
-        maxdp = rfvec[2];
-        topf = rf;
-      }
-    }
-    List_Delete(rfaces);
+      upcen[k] += fxyz[j][k];
+  }
 
-    if (!topf)
-      MSTK_Report("FixColumnPartitions","Could not find top face?",MSTK_FATAL);
+  MF_Coords(*dn,&nfv,fxyz);
+  dncen[0] = dncen[1] = dncen[2] = 0.;
+  for (j = 0; j < nfv; j++) {
+    for (k = 0; k < 3; k++)
+      dncen[k] += fxyz[j][k];
+  }
 
-    /* Check if the top face is also a top boundary face */
+  /* ASSERT x,y are the same */
+  if (fabs(upcen[0] - dncen[0]) > 1.e-12 ||
+      fabs(upcen[1] - dncen[1]) > 1.e-12) {
+    MSTK_Report("FixColumnPartitions","Mesh is not quite columnar, up and down faces do not align.",MSTK_FATAL);
+  }
 
-    fregs = MF_Regions(topf);
-    nfr = List_Num_Entries(fregs);
+  if (upcen[2] < dncen[2] - 1.e-12) {
+    rf = *up;
+    *up = *dn;
+    *dn = rf;
+    ret = 1;
+  } else if (upcen[2] > dncen[2] + 1.e-12) {
+    ret = 1;
+  } else {
+    /* find the first region upward that is proper */
+    MSTK_Report("FixColumnPartitions","Partitioning with degenerate cells is not yet supported.",MSTK_FATAL);
+    ret = 0;
+  }
 
-    if (nfr != 1) {
-      List_Delete(fregs);
-      continue; /* Not a mesh region at the top of the mesh */
-    }
-    
-    /* Hopefully, found the top face of a column in the mesh */
+  List_Delete(rfaces);
+  return ret;
+}
 
-    curreg = List_Entry(fregs,0);
-    curid = MR_ID(curreg);
-    homepid = part[curid-1];
+  
+int FixColumnPartitions(Mesh_ptr mesh, int *part, MSTK_Comm comm) { 
+  int idx, curid, nxtid, homepid, nxtpid, num_cols, num_cells_in_col;
+  int num_touches, num_not_touched;
+  int interior, done, nmove=0;
+  MRegion_ptr mr, curreg, nxtreg;
+  MFace_ptr topf, botf, topf2;
+  List_ptr fregs;
+  int* touched;
+  FILE* fid;
+  
+  if (MESH_Num_Regions(mesh) == 0) return 1; /* Not implemented for 2D (perhaps not needed either) */
+
+  /* ensure every region is touched once */
+  touched = calloc(MESH_Num_Regions(mesh),sizeof(int));
+  for (idx=0; idx!=MESH_Num_Regions(mesh); ++idx) touched[idx] = 0;
+  printf("Read %-d regions", MESH_Num_Regions(mesh));
 
+  /* loop over all regions, finding regions whose "up" face is a boundary */
+  fid = fopen("col_counts.txt", "w");
+
+  num_cols = 0;
+  num_touches = 0;
+  idx = 0;
+  while ((mr = MESH_Next_Region(mesh,&idx))) {
+    FixColumnPartitions_UpDown(mesh, mr, &topf, &botf);
+
```

Comment:
Actually, the FixColumnPartitions_UpDown is going to throw an error if either of top or botf is NULL:

`  /* find two faces that are not sides */
  found = 0; i = 0;
  while (found < 2) {
    rf = List_Entry(rfaces,i);
    if (!FixColumnPartitions_IsSide(mesh, mr, rf)) {
      if (found < 1) *up = rf;
      else *dn = rf;
      found++;
    }
    i++;
    if (i >= nrf) 
      MSTK_Report("FixColumnPartitions","Mesh is not columnar, can't find both up and down faces.",MSTK_FATAL);
  }`

This also implies that we _have_ to get both, i.e. the function will throw an error if any part of domain can be represented as columns.